### PR TITLE
Support for AND, OR, NOT filters

### LINKED ISSUE #87: Support "and", "or", "not" on where/filterBy query
**Describe the bug**
Support "and", "or", "not" on where/filterBy query

**Expected behavior**
based on hasura
Screen Shot 2022-02-21 at 5 05 20 PM

### PR DESCRIPTION
**Pull request by imor**
## What kind of change does this PR introduce?

New feature and docs update

## What is the current behavior?

Users can't combine multiple bool expression using AND/OR/NOT in queries and mutations.

## What is the new behavior?

Users will now be able to use AND/OR/NOT to combine multiple bool expressions in queries and mutations. AND and OR filters accept a list of filters while NOT accepts a single filter. For example:

```graphql
# AND filter
{
  accountCollection(
    filter: {
        AND: [
            { id: { eq: 1 } },
            { email: { eq: "aardvark@x.com" } }
        ]
    }
  ) {
    edges {
      node {
        id
        email
      }
    }
  }
}

# OR filter
{
  accountCollection(
    filter: {
        OR: [
            { id: { eq: 3 } },
            { email: { eq: "elephant@x.com" } }
        ]
    }
  ) {
    edges {
      node {
        id
        email
      }
    }
  }
}

# NOT filter
{
  accountCollection(filter:{
      NOT: { id: { eq: 3 } }
  }) {
    edges {
      node {
        id
        email
      }
    }
  }
}
```

These filters can be arbitrarily nested inside one another:

```graphql
# NOT filter nested inside OR which is nested inside AND
{
  accountCollection(
    filter: {
      AND: [
        { id: { gt: 0 } }
        { id: { lt: 4 } }
        {
          OR: [
            { NOT: { email: { eq: "bat@x.com" } } }
            { email: { eq: "cat@x.com" } }
          ]
        }
      ]
    }
  ) {
    edges {
      node {
        id
        email
      }
    }
  }
}
```

Empty filters will be ignored:

```graphql
# The following two queries are equivalent
{
  accountCollection(filter: { AND: [], OR: [], NOT: {} }) {
    edges {
      node {
        id
        email
      }
    }
  }
}

{
  accountCollection {
    edges {
      node {
        id
        email
      }
    }
  }
}

```

Multiple simple filters at any depth level will be implicitly ANDed, this is exactly like how simple filters work today:

```graphql
# Filter will compile to NOT(id = 1 AND email = 'aardvark@x.com')
{
  accountCollection(
    filter: {
        NOT: {
            # id and email filters implicitly ANDed
            id: { eq: 1 },
            email: { eq: "aardvark@x.com" }
        }
    }
  ) {
    edges {
      node {
        id
        email
      }
    }
  }
}

```

Update and delete mutations will also accept these filters.

If a column name (inflected) is `AND`, `OR`, or `NOT` then that compound filter will be disabled. Users should either not name their columns with these values or rename them using a comment directive to avoid confusing queries.

For full details of the behaviour see the new tests added in this PR.

## Additional context

- [X] Support AND and OR filters
- [x] Support NOT filters
- [x] Write exhaustive tests
- [x] Last minute cleanup/clippy fixes etc.
- [x] Update docs

Resolves #87

### CODE REVIEW

**Comment by imor** [comment]
@olirice could you please do an initial review of the design before I start working on the docs. The design is quite similar to [Hasura's design](https://hasura.io/docs/latest/api-reference/graphql-api/query/#whereexp). The difference is in naming, they name the filters `_and`, `_or` and `_not`.

**Comment by imor** [comment]
Wrote docs as well, marking this PR ready for review.

**Comment by olirice** [comment]
> The major difference is in naming, they name the filters _and, _or and _not and ours are named AND/OR/NOT.
>
>I'm thinking of lowercasing the operator names from AND/OR/NOT to and/or/not. This will make them consistent with >operators like eq/neq etc. What do you think?

We discussed it internally. Hasura uses `_and`, prisma uses `AND` and postgraphile doesn't support or by default so it auto-composes filters with and

the original concern with using `and` is that it could conflict with column names but that seems unlikely and using lowercase for consistency would be nice.

yes, please convert to lower case. thanks!

**Comment by imor** [comment]
> the original concern with using and is that it could conflict with column names but that seems unlikely

It is a valid concern and I think Hasura and Prisma name their operators to avoid conflicts with column names. Can we maybe collect real world data about column names to find out what are the chances of collision? If that is not possible maybe a poll from the community can give us an idea.

> using lowercase for consistency would be nice.

For now I've lowercased the names.

File: src/graphql.rs [hunk 1]
```diff
@@ -3440,6 +3458,74 @@ impl ___Type for FilterEntityType {
...
+        // during filter argument validation both of these will be checked
+        // against any key in the input named `and` and one of them will fail
+        // the validation. This means the user can neither use a simple filter
+        // on the `and` column nor use `and` in compound expressions.
+        //
+        // To prevent this we do not add the `and` filter entry which essentially
+        // disables the `and` compound filter. But at least the user is able to
+        // use simple column filters.
+        //
+        // Similiar logic applies for `or` and `not` filters.
```
Comment (olirice) [inline src/graphql.rs:3473]: love this, nicely done

**Review by olirice — APPROVED** [review]

**Comment by olirice** [comment]
From our publicly visible projects, there are no existing collisions. With the logic you put in place to skip `and`/`or`/`not` when a column with the same name exists, I think we're good to use the lower case version

awesome work!

**Pull request merged**